subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/verification/DependencyVerifierBuilder.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.api.internal.artifacts.verification;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Maps;
+import org.gradle.api.artifacts.component.ComponentIdentifier;
+import org.gradle.api.artifacts.component.ModuleComponentIdentifier;
+import org.gradle.api.internal.artifacts.verification.model.ArtifactVerificationMetadata;
+import org.gradle.api.internal.artifacts.verification.model.ChecksumKind;
+import org.gradle.api.internal.artifacts.verification.model.ComponentVerificationMetadata;
+import org.gradle.api.internal.artifacts.verification.model.ImmutableArtifactVerificationMetadata;
+import org.gradle.api.internal.artifacts.verification.model.ImmutableComponentVerificationMetadata;
+import org.gradle.internal.component.external.model.ModuleComponentArtifactIdentifier;
+
+import java.util.EnumMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class DependencyVerifierBuilder {
+    private final Map<ModuleComponentIdentifier, ComponentVerificationsBuilder> byComponent = Maps.newLinkedHashMap();
+
+    public void addChecksum(ModuleComponentArtifactIdentifier artifact, ChecksumKind kind, String value) {
+        ModuleComponentIdentifier componentIdentifier = artifact.getComponentIdentifier();
+        byComponent.computeIfAbsent(componentIdentifier, id -> new ComponentVerificationsBuilder(id))
+            .addChecksum(artifact, kind, value);
+    }
+
+    public DependencyVerifier build() {
+        ImmutableMap.Builder<ComponentIdentifier, ComponentVerificationMetadata> builder = ImmutableMap.builderWithExpectedSize(byComponent.size());
```

Comment:
Are you not loosing the order here?

Reply:
No, `ImmutableMap.Builder` guarantees same ordering.